frontend/src/components/widget/WidgetRenderer.js

```diff
@@ -0,0 +1,629 @@
+          renderContent = <textarea {...widgetProperties} />;
+        }
+        return (
+          <div>
+            <div
+              className={classnames(this.getClassNames(classNameParams), {
+                'border-danger': showErrorBorder,
+              })}
+            >
+              {renderContent}
+            </div>
+            {charsTyped && charsTyped[fieldName] >= 0 && (
+              <CharacterLimitInfo
+                charsTyped={charsTyped[fieldName]}
+                maxLength={maxLength}
+              />
+            )}
+          </div>
+        );
+      }
+      case 'Password':
+        return (
+          <div className="input-inner-container">
+            <div
+              className={classnames(
+                this.getClassNames({
+                  icon: true,
+                })
+              )}
+            >
+              <input {...widgetProperties} type="password" />
+              {icon && <i className="meta-icon-edit input-icon-right" />}
+            </div>
+            {allowShowPassword && (
+              <div
+                onMouseDown={() => onSetWidgetType('text')}
+                onMouseUp={() => onSetWidgetType('password')}
+                className="btn btn-icon btn-meta-outline-secondary btn-inline pointer btn-distance-rev btn-sm"
+              >
+                <i className="meta-icon-show" />
+              </div>
+            )}
+          </div>
+        );
+      case 'Integer':
+      case 'Amount':
+      case 'Quantity':
+        return (
+          <div className={classnames(this.getClassNames(), 'number-field')}>
+            <input
+              {...widgetProperties}
+              type="number"
+              min={0}
+              precision={widgetField === 'CableLength' ? 2 : 1}
+              step={subentity === 'quickInput' ? 'any' : 1}
+            />
+            {widgetData[0].devices && (
+              <div className="device-widget-wrapper">
+                <DevicesWidget
+                  devices={widgetData[0].devices}
+                  tabIndex={1}
+                  handleChange={(value) =>
+                    onPatch && onPatch(fields[0].field, value)
+                  }
+                />
+              </div>
+            )}
+          </div>
+        );
+      case 'Number':
+      case 'CostPrice':
+        return (
+          <div className={classnames(this.getClassNames(), 'number-field')}>
+            <input {...widgetProperties} type="number" />
+          </div>
+        );
+      case 'YesNo':
+        return (
+          <Checkbox
+            {...{
+              widgetData,
+              disabled: readonly,
+              fullScreen,
+              tabIndex,
+              widgetField,
+              id,
+              filterWidget,
+            }}
+            handlePatch={onPatch}
+          />
+        );
+      case 'Switch':
+        return (
+          <label
+            className={classnames('input-switch', {
```

Comment:
Basically if we reduce the size of the file and have the logic in separate granular components like in here <Switch {...props} /> I believe logic would be easy to follow and any further code changes would be smoother.

Reply:
Well yes and no. It would be fairly straightforward if all the components would take the same props. But they don't. So you either have to pass all the props, that `WidgetRenderer` is getting, or build props object for each `widgetType`, or just define props in the element definition (so pretty much the same thing we have right now). So only in first case this would make the code shorter, but then every widget would get bazillion unnecessary properties that it won't use in any way. In the other two cases, we would only replace `<label>whatever</label>` with a component name.